kernel/src/common/regs/README.md

```diff
@@ -0,0 +1,207 @@
+# Register/Bitfield Interface for Tock
+
+This module provides an interface for defining and manipulating memory mapped
+registers and bitfields.
+
+## Defining registers
+
+The module provides three types for working with memory mapped registers: `ReadWrite`,
+`ReadOnly`, and `WriteOnly`, providing read-write, read-only, and write-only functionality,
+respectively.
+
+Defining the registers is similar to the C-style approach, where each register
+is a field in a packed struct:
+
+```rust
+use common::regs::{ReadOnly, ReadWrite, WriteOnly};
+
+#[repr(C, packed)]
+struct Registers {
+    // Control register: read-write
+    // The 'Control' parameter constrains this register to only use fields from
+    // a certain group (defined below in the bitfields section).
+    cr: ReadWrite<u8, Control>,
+
+    // Status register: read-only
+    s: ReadOnly<u8, Status>
+
+    // Registers can be bytes, halfwords, or words:
+    // Note that the second type parameter can be omitted, meaning that there
+    // are no bitfields defined for these registers.
+    byte0: ReadWrite<u8>,
+    byte1: ReadWrite<u8>,
+    short: ReadWrite<u16>,
+    word: ReadWrite<u32>
+
+    // Etc.
+}
+```
+
+## Defining bitfields
+
+Bitfields are defined through the `bitfields!` macro:
+
+```rust
+bitfields! [
+    // First parameter is the register width for the bitfields. Can be u8, u16,
+    // or u32.
+    u8,
+
+    // Each subsequent parameter is a register abbreviation, its descriptive
+    // name, and its associated bitfields.
+    // The descriptive name defines this 'group' of bitfields. Only registers
+    // defined as ReadWrite<_, Control> can use these bitfields.
+    CR Control [
```

Comment:
I think for users the multiple names here is probably confusing when trying to understand how to use or read code with this register format. Are the two names (`CR` and `Control`) fundamental or could that be condensed to one?

Reply:
The two names serve different purposes. Fundamentally, the first name corresponds to a namespace, and the second name corresponds to a unique type. It's definitely possible to eliminate the second name and automatically name the type as eg CR::Type or something. But this changes the register definition from `pub cr: ReadWrite<u32, Control>` to `pub cr: ReadWrite<u32, CR::Type>`. I'm not sure which way would be more confusing.

Reply:
Sounds good, I buy it. The one thing I do like about using abbreviations for registers, though, is that it's a common convention, and it helps visually emphasize code that's messing with registers. It also tends to be less ambiguous than user-selected descriptive names. But I'm totally open to using descriptive names if that's preferred.